utils/CircuitCheck/CircuitCheck.cpp

```diff
@@ -21,98 +21,95 @@
-      llvm::errs() << "Cannot compute unitary for " << opName.str() << ".\n";
-      exitStatus = EXIT_FAILURE;
-      continue;
+    cl::ParseCommandLineOptions(argc, argv);
+
+    // Initialize MLIR context and load required dialects
+    MLIRContext context;
+    context.loadDialect<cudaq::cc::CCDialect, quake::QuakeDialect, func::FuncDialect>();
+
+    // Parse input and check files
+    ParserConfig config(&context);
+    auto checkMod = parseSourceFile<mlir::ModuleOp>(checkFilename, config);
+    auto inputMod = parseSourceFile<mlir::ModuleOp>(inputFilename, config);
+    if (!checkMod || !inputMod) {
+        llvm::errs() << "Error parsing input files.\n";
+        return EXIT_FAILURE;
     }
 
-    // Here we use std streams because Eigen printers don't work with LLVM ones.
-    if (!cudaq::isApproxEqual(checkUnitary, inputUnitary, upToGlobalPhase)) {
-      applyTolerance(checkUnitary);
-      applyTolerance(inputUnitary);
-      std::cerr << "Circuit: " << opName.str() << '\n';
-      std::cerr << "Expected:\n";
-      std::cerr << checkUnitary << '\n';
-      std::cerr << "Got:\n";
-      std::cerr << inputUnitary << '\n';
-      exitStatus = EXIT_FAILURE;
+    // Apply canonicalization if enabled
+    if (!dontCanonicalize) {
+        PassManager pm(&context);
+        OpPassManager &nestedFuncPM = pm.nest<func::FuncOp>();
+        nestedFuncPM.addPass(createCanonicalizerPass());
+        if (failed(pm.run(*checkMod)) || failed(pm.run(*inputMod))) {
+            llvm::errs() << "Canonicalization failed.\n";
+            return EXIT_FAILURE;
+        }
     }
 
-    if (printUnitary) {
-      applyTolerance(checkUnitary);
-      std::cout << "Circuit: " << opName.str() << '\n'
-                << checkUnitary << "\n\n";
+    // Apply tolerance to small numerical values in matrices
+    auto applyTolerance = [](cudaq::UnitaryBuilder::UMatrix &m) {
+        m = (1e-12 < m.array().abs()).select(m, 0.0f);
+    };
+
+    // Iterate through functions in check module and compare against input module
+    int exitStatus = EXIT_SUCCESS;
+    for (auto checkFunc : checkMod->getOps<func::FuncOp>()) {
+        StringAttr opName = checkFunc.getSymNameAttr();
+        cudaq::UnitaryBuilder::UMatrix checkUnitary, inputUnitary;
+
+        // Look up corresponding function in input module
+        auto *inputOp = inputMod->lookupSymbol(opName);
+        if (!inputOp) {
+            llvm::errs() << "Function " << opName.str() << " not found in input.\n";
+            exitStatus = EXIT_FAILURE;
+            continue;
+        }
+
+        auto inputFunc = dyn_cast<func::FuncOp>(inputOp);
+        if (!inputFunc || failed(computeUnitary(checkFunc, checkUnitary)) ||
+            failed(computeUnitary(inputFunc, inputUnitary, upToMapping))) {
+            llvm::errs() << "Cannot compute unitary for " << opName.str() << ".\n";
+            exitStatus = EXIT_FAILURE;
+            continue;
+        }
```

Comment:
Since you are already improving this by checking `inputFunc`, it would be good to go one step further and be explicit about what exactly failed, e.g., was the computation of `inputFunc`'s unitary or `checkFunc`'s one?